File: autoingestion.rb
require 'util/http_verify_none.rb'
require 'net/https'
require 'uri'
require 'cgi'

def main
  autoingestor = Autoingestion.new
  yesterday = (Time.new - 24*60*60).strftime("%Y%m%d")

  if File.exists?("config.hash")
    config = eval(File.new("config.hash").read())
    autoingestor.username = config["username"]
    autoingestor.password = config["password"]
    autoingestor.vendor_number = config["vendor_number"]
  else
    print "iTunes Connect email address:             "
    autoingestor.username = gets.strip!
    print "iTunes Connect password:                  "
    autoingestor.password = gets.strip!
    print "Vendor number:                            "
    autoingestor.vendor_number = gets.strip!
    print "Type of report:                [Sales]    "
    autoingestor.type_of_report = gets.strip!
    print "Date type:                     [Daily]    "
    autoingestor.date_type = gets.strip!
    print "Report type:                   [Summary]  "
    autoingestor.report_type = gets.strip!
    print "Report date:                   [#{yesterday}] "
    autoingestor.report_date = gets.strip!
  end

  autoingestor.type_of_report = "Sales"   if autoingestor.type_of_report.empty?
  autoingestor.date_type      = "Daily"   if autoingestor.date_type.empty?
  autoingestor.report_type    = "Summary" if autoingestor.report_type.empty?
  autoingestor.report_date    = yesterday if autoingestor.report_date.empty?

  autoingestor.perform_request
end

class Autoingestion

  ITUNES_URL = "https://reportingitc.apple.com/autoingestion.tft"

  attr_accessor :username
  attr_accessor :password
  attr_accessor :vendor_number
  attr_accessor :type_of_report
  attr_accessor :date_type
  attr_accessor :report_type
  attr_accessor :report_date

  def create_post_body
    data  = "USERNAME="     + CGI.escape(@username.to_s)       + "&"
    data += "PASSWORD="     + CGI.escape(@password.to_s)       + "&"
    data += "VNDNUMBER="    + CGI.escape(@vendor_number.to_s)  + "&"
    data += "TYPEOFREPORT=" + CGI.escape(@type_of_report.to_s) + "&"
    data += "DATETYPE="     + CGI.escape(@date_type.to_s)      + "&"
    data += "REPORTTYPE="   + CGI.escape(@report_type.to_s)    + "&"
    data += "REPORTDATE="   + CGI.escape(@report_date.to_s)
  end

  def perform_request
    uri = URI.parse(ITUNES_URL)
    
    request = Net::HTTP.new(uri.host, uri.port)
    request.use_ssl = true
    
    body = create_post_body
    headers = {'Content-Type' => 'application/x-www-form-urlencoded'}
    response = request.post2(uri.path, body, headers)

    if response['filename'] != nil
      filename = response['filename']
      f = File.new(filename, "w")
      f.write(response.body)
      f.close
      puts "File Downloaded Successfully (#{filename})"

      system("cp #{filename} #{filename}.bak")
      system("gunzip #{filename}")
      system("mv #{filename}.bak #{filename}")
      
    elsif response['errormsg'] != nil
      puts response['errormsg']
    else
      puts "No recognized response, dumping headers.."
      response.each_header do | key, value |
        puts "#{key}: #{value}"
      end
    end
  end

end

if __FILE__ == $PROGRAM_NAME
  main
end

File: Autoingestion.java
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintStream;

import java.net.URL;
import java.net.URLEncoder;
import java.net.HttpURLConnection;

import javax.net.ssl.HttpsURLConnection;

import java.text.SimpleDateFormat;
import java.util.Calendar;

/**
 * This is the class provided by Apple for downloading iTunes Connect reports.<br />
 * Decompiled with JD-GUI and cleaned up by William LaFrance.<br />
 * <br />
 * Find more information in section 3.2 of the iTunes Connect Sales and Trends Guide.<br />
 * http://www.apple.com/itunesnews/docs/AppStoreReportingInstructions.pdf
 */
public class Autoingestion
{

    /**
     * Command-line entry point<br />
     * <br />
     * Usage: <i>java Autoingestion john@xyz.com letmein 80012345 Sales Daily Summary 20100204</i>
     * <br /><br />
     * @param arguments	Command line arguments
     */
    public static void main(String[] arguments) throws Throwable
    {
        if ((arguments.length < 6) || (arguments.length > 7)) {
            System.out.println("Please enter all the required parameters. For help, please "
                    + "download the latest User Guide from the Sales and Trends module in iTunes "
                    + "Connect.");
            return;
        }

        String date = null;

        if ((arguments.length == 7) && (arguments[6]) != null) {
            date = arguments[6];
        } else {
            Calendar calendar = Calendar.getInstance();
            SimpleDateFormat formatter = new SimpleDateFormat("yyyyMMdd");
            calendar.add(5, -1);

            date = formatter.format(calendar.getTime()).toString();
        }

        String data = "USERNAME="      + URLEncoder.encode(arguments[0], "UTF-8");
        data = data + "&PASSWORD="     + URLEncoder.encode(arguments[1], "UTF-8");
        data = data + "&VNDNUMBER="    + URLEncoder.encode(arguments[2], "UTF-8");
        data = data + "&TYPEOFREPORT=" + URLEncoder.encode(arguments[3], "UTF-8");
        data = data + "&DATETYPE="     + URLEncoder.encode(arguments[4], "UTF-8");
        data = data + "&REPORTTYPE="   + URLEncoder.encode(arguments[5], "UTF-8");
        data = data + "&REPORTDATE="   + URLEncoder.encode(date, "UTF-8");
    
        HttpURLConnection connection = null;
        try {
            URL url = new URL("https://reportingitc.apple.com/autoingestion.tft?");

            connection = (HttpURLConnection)url.openConnection();
      		connection.setRequestMethod("POST");
      		connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
      		connection.setDoOutput(true);
      		
      		OutputStreamWriter out = new OutputStreamWriter(connection.getOutputStream());
      		out.write(data);
      		out.flush();
      		out.close();

      		if (connection.getHeaderField("ERRORMSG") != null) {
        		System.out.println(connection.getHeaderField("ERRORMSG"));
      		} else if (connection.getHeaderField("filename") != null) {
        		getFile(connection);
        	}
    	} catch (Exception ex) {
      		ex.printStackTrace();
      		System.out.println("The report you requested is not available at this time.  Please try again in a few minutes.");
    	} finally {
      		if (connection != null) {
        		connection.disconnect();
        		connection = null;
      		}
    	}
    }

    /**
     * Download a file from a HTTP connection and write it to file indicated by
     * HTTP header 'filename'
     *
     * @param connection HTTP connection to read downloading file from
     *
     * (Note: The argument was HttpsURLConnection in decompiled code)
     */
	private static void getFile(HttpURLConnection connection) throws IOException {
    	String filename = connection.getHeaderField("filename");
    	System.out.println(filename);

    	BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
    	BufferedOutputStream out = new BufferedOutputStream(new FileOutputStream(filename));

    	int i = 0;
    	byte[] data = new byte[1024];
    	while ((i = in.read(data)) != -1) {
      		out.write(data, 0, i);
    	}

    	in.close();
    	out.close();
    	System.out.println("File Downloaded Successfully ");
  	}
}
